### DISCUSSION TITLE
Apple II VGA PCB

### DISCUSSION THREAD

**Discussion by swetland** [markadev/AppleII-VGA#15] [Show and tell]
I recently bought an Apple IIe on ebay and was thinking "maybe you could do a video card by just snooping bus writes"... a bit of searching trying to figure out if that was feasible landed me here where I got my answer along with "and it's already been done."

I thought it'd be fun to build one and I'm not fond of working with perfboard so I fired up KiCAD and ended up with this...

![appleii vga schematic v1 0 2](https://github.com/markadev/AppleII-VGA/assets/94628/dfd08655-9b56-4cce-9ba4-343eff228184)

![appleii vga render v1 0 2](https://github.com/markadev/AppleII-VGA/assets/94628/0f577d78-51bf-4459-b103-04652c80c3bb)

![appleii vga layout v1 0 2](https://github.com/markadev/AppleII-VGA/assets/94628/57111a48-a1a6-436a-9cbf-9c2820645dd3)

I kept the design very close to the original, so I could try the firmware with minimal changes.  I used four buffers for all the bus signals to reduce the number of IC types needed, and the board can be configured to make D0..D7 input only and route SYNC to the Pico instead of direction control for D0..D7 as an option to explore trying to synchronize video updates.

**Comment by swetland** [comment]
My PCBs arrived from JLCPCB today (cost worked out to about $5.10 ea for the card and a stub for the rear panel, for 5 boards) and parts had arrived earlier this week, so time to put them together and see if things worked...

![appleiivga-pcb-1](https://github.com/markadev/AppleII-VGA/assets/94628/2e6fb4c6-e3e3-447a-b974-618943005c98)

![appleiivga-pcb-2](https://github.com/markadev/AppleII-VGA/assets/94628/1b8616eb-e3a3-4ac0-9b36-bbb040025246)

![appleiivga-assembled-1](https://github.com/markadev/AppleII-VGA/assets/94628/0d5ad359-75ae-400a-acfa-885c80c36e78)

![appleiivga-test-pattern-sony-crt](https://github.com/markadev/AppleII-VGA/assets/94628/e733a9ab-4d08-4147-803e-51280fdd5bc0)

Two changes I'd make for the next revision: Fix the "1uF" labels to read "0.1uF" as they should, and make the card fingers section maybe 1-2mm narrower (the boards are rather snug in the expansion socket).

**Reply by techsurrusr** [reply]
Curious but is there anywhere once can order this card since having the vga output as you've designed this is far better than some of the similar card design?

**Reply by markadev** [reply]
> Curious but is there anywhere once can order this card since having the vga output as you've designed this is far better than some of the similar card design?

Hey there, I've seen some inexpensive cards going on ebay that have the separated VGA connector.

**Comment by swetland** [comment]
I disabled the read cycle part of abus.c and abus.pio and tried it out in the IIe with very positive results:

![appleiivga-test-1](https://github.com/markadev/AppleII-VGA/assets/94628/318b2f95-602b-4742-9f7c-b94d1c066aa0)

![appleiivga-test-2](https://github.com/markadev/AppleII-VGA/assets/94628/709c42d0-5bb6-45da-9da7-fd3913396df8)

![appleiivga-test-3](https://github.com/markadev/AppleII-VGA/assets/94628/ce994d4c-011d-4c17-a22d-baed75c523c5)

![appleiivga-test-4](https://github.com/markadev/AppleII-VGA/assets/94628/ae131514-060a-426c-aa28-e4c3e96936c2)

I'm planning on tidying stuff up and pushing the project to github soon (under some permissive license) and also thinking about doing a version with a TI TFP410 DVI/HDMI encoder as well -- could provide both VGA and HDMI from a single board, I believe.

**Comment by markadev** [comment]
Nice board, the output looks good.  Thanks for sharing!

Those are some good trade-offs you made in the design for a more IIe focused board.  I originally wanted to add Videx emulation for II+ 80-column support but not much software really used it so ultimately abandoned the idea.  Without that there's not a good reason to keep the device read_cycle support.

Did I read that right, that you paid ~$26 for 5 PCBs with gold-plated edge connectors? Wow, I would've also taken that deal instead of hand wiring 😀

**Reply by aaronsmithdev** [reply]
I'd love to see Videx emulation for the ][+! I regularly use some terminal programs, a modified version of Appleworks for the ][+, and a handful of custom basic programs that need 80 columns. I have your board and a Videx in my machine and have started poking around the Videx manual for memory addresses. I was going to start toying with the code soon. I keep my composite monitor hooked up now only because that's the only way I can get 80 columns at the moment. Mark, if you have any notes or tips from when you were considering this, I'd really appreciate seeing them! Regardless, I'll come back and post my code when I've had time to sit down with it.

I've added lower case characters for the ][+ using the Videx lowercase chip ROM code already and plan on posting that as well. You would think it would be of limited value, but with all the cracked programs out there with custom title and option screens (I'm assuming created on a //e without thought of backwards compatibility), it's been really great. If I can get Videx emulation working, it'll pair nicely with the lower case character set.

Thanks for the great card and all the hard work!

**Reply by markadev** [reply]
Hey there, welcome! Videx support would be pretty cool to have especially if you have software that uses it.

Originally, I had the idea to fully emulate a Videx card on the Pico, including all the onboard ROM, RAM, and registers, so that the VGA card would fully replace the Videx card.  I think this is probably doable but I'm not entirely sure if the current hardware design can support it.

The read_cycle support mentioned above was just the PIO and C code used to respond to CPU reads (by switching the IO pin directions and putting data on the data bus) and something like it would be needed for a fully emulated Videx card.  The original code is [still in the git history](https://github.com/markadev/AppleII-VGA/blob/read-cycle-last-call/pico/abus.pio#L44-L104) but it's not nearly flexible enough to support full Videx emulation and was just a proof-of-concept.

But, while fully emulating a Videx card would be nice, there could be a simpler solution that just adds "Videx compatibility" by inspecting all the bus cycles to slot 3, reconstructing the Videx video memory, and rendering the output.  The upside is that the Pico would just be passively inspecting the bus just like it does for all the other Apple video modes so it would work similarly with no limitations imposed by the current hardware design.  The downside of course is that you'd still need a real Videx with its ROMs and video RAM.  And I guess to be fair, 80 column mode on an Apple IIe still requires the Apple IIe 80 column card to be present (to host the video RAM) so it's at parity.

Hah, so now that I've organized my thoughts and typed this all out it seems like "Videx compatibility" is something that should definitely be added if that's something you'd like to try.  It's not a perfect solution but could solve the 80-column problem for a lot of people and avoids the complexity that full Videx emulation would add.  Or, feel free to look at full Videx emulation if you'd like to - there would be a whole set of other topics that I didn't mention 😀

**Reply by aaronsmithdev** [reply]
Awesome - thanks, Mark! That's a great call: watch for the traffic instead of trying to emulate it. Since I have the card, I'll definitely stay away from the emulation. I had started doing some research on what it would take and it made my head spin. As I understand it, Videx was one of the first (if not first) 80-column cards on the market for the II and II+ and others basically started to standardize around what they had done, so that might make whatever changes may come of this more broadly applicable.

I appreciate the guidance! I see another 80-column post was opened up, so I'll also try to keep any conversation from here-on-out in the right thread. I'll probably sit down and see what I can do in the next couple of weeks.

By-the-way, I've created a fork and put my lowercase modifications into it, along with a snazzy configuration utility disk. I created a pull request for the lowercase modifications (although I'm not sure if you want it in main since you provided the python script that was needed to do that). I shied away from creating one for the config disk since I'm not entirely sure if I'm missing some things for the IIe and couldn't test (alas, I only have the II+ at the moment).

Much appreciated and I hope you have a great New Year!

![PXL_20231230_023211095](https://github.com/markadev/AppleII-VGA/assets/82830826/09be811c-a10d-4e92-a4dd-5c6042a7c512)

**Reply by abaffa** [reply]
Thanks @markadev for suggesting this thread to me :) . 

@aaronsmithdev was wondering if iie 80 columns mode could be adapted to offer 80 columns support to apple ii+ machines. I read once that apple based the iie 80 col mode on videx.

I've been looking at the videx addresses and trying to understand how to emulate it.  I'll let you know when I get results.

Best
Augusto Baffa

**Reply by abaffa** [reply]
I've just found (by chance) an [videx emulator code written in C](https://mirrors.apple2.org.za/Apple%20II%20Documentation%20Project/Interface%20Cards/80%20Column%20Cards/Videx%20Videoterm/Documentation/Videx%20Videoterm%20-%20Emulation.c) ... and also the [schematics](https://mirrors.apple2.org.za/Apple%20II%20Documentation%20Project/Interface%20Cards/80%20Column%20Cards/Videx%20Videoterm/Schematics/)

**Reply by aaronsmithdev** [reply]
Hi @abaffa - I candidly haven't gotten much of a start on this myself but was definitely wondering the same and wanted to take a harder look into it. I looked over Mark's code and saw how it might be implemented and have the Videx as well. It was also my understanding that the Apple IIe 80 column mode was based on the Videx design, so I'm also assuming if the VGA card can handle it, it won't be much different than what's been implemented for the IIe mode.

I hadn't run across that C code for the Videx emulator, though, so that is very helpful! I use the 80-column mode on my II+ frequently and would love to be able to use it with the card.

Glad to hear there are others wanting to see this implemented! I'll check in with you if I happen to get anywhere with it and will be eagerly awaiting your results if you happen to move forward more quickly - maybe we can get farther along together!

**Reply by abaffa** [reply]
Hi There, I almost understood how videx works. I managed to implement the roms and part of the framebuffer in my emulator program. There are few bugs but when executing pr#3, I can enter 80 columns mode and also I can exit it by typing ctrl+Z+1. But there are still some states that I am not understanding for example, I can't clear the screen with the "home" command and I haven't yet identified which signal the card should read. I'll try running the demo disc to see if I can get some clues :)

All code is written in C language.
![80columns](https://github.com/markadev/AppleII-VGA/assets/7242662/09316ee5-e87f-4b22-b2a9-4b77e1c67547)

**Reply by abaffa** [reply]
well... after reading the manual, I understood that some basic commands do not work: home, inverse, flash should be ignored. To erase the screen you need to type Ctrl+L. I also added the "inverse" character rom as a secondary rom and could swap it with CTRL+Z+2 and CTRL+Z+3. The only function that doesn't work and I still need to understand is CTRL+A (smallcase/upcase). Other than that, I have all the videx logic implemented. Unfortunately the original videx ROM does not support the full integration of basic as was done in apple//e but it is the best option for appleII+ users.

It would be necessary to spend just over 6KBs to have the firmware + character rom in the RPI source code. For RAM usage: 2K for VRam and 17 bytes for registers. It's a new video mode in addition to the already implemented text (with page1/page2), gr, hgr etc...

It will be the salvation for many appleiiplus lovers :)... I really want to put an 80 column card in my old one...

**Reply by abaffa** [reply]
All done! The code is tested and running on my emulator. Now I (or we) need to figure how to insert it to AppleIIVGA project. My code (notes and comments) [here](https://gist.github.com/abaffa/c92ee8898318d4043b19e7aadc1b5e46) for further reference. 

PS: It uses the videx firmware and two char roms in another file. It's necessary to download them [here](https://mirrors.apple2.org.za/Apple%20II%20Documentation%20Project/Interface%20Cards/80%20Column%20Cards/Videx%20Videoterm/ROM%20Images/).

![image](https://github.com/markadev/AppleII-VGA/assets/7242662/e2c0073b-9fff-4b1f-a60f-6ed74e9ef1ee)

**Reply by abaffa** [reply]
I found an issue: we must revert the code to REV A in order to use GPIO10 as IN/OUT control to data.
I was having some progress but I got confused about how to write to bus during IOSEL/DEVSEL/IOSTB. It's necessary to read the firmwware/rom from the board and also the special memory page from videx.

@markadev @aaronsmithdev can you help me? if we're able to read data from the card I can finish it :)

// abus.c
//
// In the rev A schematic this pin was originally used to control the data bus pins transceiver direction
// so that bus reads could be responded to with data. This code has since been removed so the GPIO could be
// repurposed.
//
// A pull-down is set on this pin to remain compatible with these rev A based designs. This will ensure that
// by default the data transceiver direction in "inward".
  gpio_set_pulls(CONFIG_PIN_APPLEBUS_SYNC, false, true);

**Reply by abaffa** [reply]
Good news... I read how to implement i/o from the V2Analog project. After a day working with timings, the apple2 is finally reading the firmware and responding when I type some trash. All data is stored in videx memory and I could revert the PR#3 using ctrl+z+1. 

Now I need to adapt the vga rendering code to read the memory from videx memory array and draw the characters. This is the easy part.

**Reply by abaffa** [reply]
it's almost there... few bugs but running https://youtu.be/gcQpmpK4G-8
The bus timing is tricky sometimes.

**Reply by abaffa** [reply]
finishing this thread - all done. I found out the videx card is not compatible to grappler+ (printer interface) 'cos they share the same banking rom solution using I/O STB address. I had to remove my printer interface to have the videx running :D. Also, I decided to map slot 3 directly from the memory address in order to leave the devsel signal to change the character rom fonts. So the card can be used in slot 7 but it always maps videx card to slot 3. (Let slot 3 empty and the card in slot 7 will answer to both slots)

The code is available at https://github.com/abaffa/AppleII-VGA/ and I requested a pull https://github.com/markadev/AppleII-VGA/pull/38 so Mark can verify my solution and suggest some project standards.

**80 Columns mode on AppleII+ - Videos**

[![AppleII Videx Videoterm 80 Columns card Demo running at AppleIIVga by Markadev](http://img.youtube.com/vi/jKAShbG-p9Y/0.jpg)](https://www.youtube.com/watch?v=jKAShbG-p9Y "Apple][+ Videx Videoterm 80 Columns card Demo running at AppleIIVga by Markadev")

[![AppleII Accessing BBS using Videx Videoterm 80 Columns card running at AppleIIVga by Markadev](http://img.youtube.com/vi/D4BjUacbDPo/0.jpg)](https://www.youtube.com/watch?v=D4BjUacbDPo "Apple][+ Accessing BBS using Videx Videoterm 80 Columns card running at AppleIIVga by Markade")

**Comment by swetland** [comment]
I'm interested in playing with the read cycle stuff to maybe implement some kind of IO, network, or disk interface, though I think I'd probably want /IOSEL as well as /DEVSEL for that, so a slightly different board might be called for.

JLCPCB is rather impressively inexpensive and fast -- $23.50 for 5 of the main PCBs and $2 for 5 of the little VGA connector + header board (HASL not ENIG for those) for the rear panel.  $17 for shipping.  Six days from placing the order to boards arriving by DHL.

I love the "use multiple buffers to sample chunks of the bus in sequence" approach to conserving GPIOs and I am absolutely going to steal the idea for a ZX81 IO interface project to simplify things there -- there's a little less headroom at 3.5MHz, but I think it should still be workable (and I probably only need two buffers of capture, worst case, for that one).

I also changed the power setup slightly -- using the 3V3 out from the Pico to power the buffers and supplying 5V to the Pico's VSYS from the Apple II bus, using a P-FET to disable this if VBUS is present (per the suggestion in section 4.5 of the Pico datasheet).

**Reply by markadev** [reply]
Cool, yeah I think there's a lot of potential using cheap/fast modern microcontrollers like the Pico to build new or emulate older peripherals, like [this project](https://github.com/V2RetroComputing/analog-firmware) that emulates a full Z80 Applicard.

Fair warning that the current read_cycle implementation is not that good because the ARM core is just a little bit too slow to get the data into the PIO's FIFO in time reliably.  A couple solutions I've seen are to overclock the Pico, or [this guy's fascinating project](https://github.com/picocomputer/rp6502) where the PIO program emulating a ROM chip directly triggers a DMA engine & bypasses the CPU completely.  There's definitely some interesting ideas to play with in this space 😁

**Comment by swetland** [comment]
I pushed everything to github here: https://github.com/swetland/appleii.vga

**Reply by markadev** [reply]
Hey Brian, I'm finally designing my own board now (with through-hole components) and you've already validated many of the changes I was thinking of making, so thanks very much for that!  I'm planning to make your SYNC->GPIO10 connection the default connection, with cuttable jumpers if I ever want to reconfigure to experiment with read-cycle support again.

Also, the separate connector board is a great idea and makes the IIe installation very neat.  It's a way better idea than what I was thinking of.

**Reply by swetland** [reply]
One thing to keep in mind is that clearance between the appleiivga board and the connector board can be a little tight.  Using right-angle connectors helps a bit, but things could probably be arranged a little better than what I landed on for my first attempt:

![appleiivga-clearance](https://github.com/markadev/AppleII-VGA/assets/94628/e3e0afa4-19ab-422c-94af-56c4ddd07237)

**Reply by markadev** [reply]
I've made some PCBs and see what you mean now 😀.  The VGA connector itself that I have is pretty deep so I'm not sure it's possible to use a right-angle connector without blocking a slot unless the pin header was below the VGA connector.

I didn't need slot 4 anyway so it works for me and I really like the clean look!  If I do another PCB run then maybe I'll try another layout.

**Reply by swetland** [reply]
With the single-row 0.1" head, a right-angle on the main PCB avoids issues with an adjacent card -- I'm not sure if the problem is avoidable with the VGA connector board -- maybe with a slightly shallower connector and making the PCB as small as possible...

**Comment by swetland** [comment]
Looking at disassembly of abus_loop(), it seems like the compiler is inlining a lot of stuff and generating a fair bit of code.  

When I was poking at my ZX81 IO project I wasn't even using the PIO, just having core1 implement the Z80 bus interface in a tight loop written in thumb assembly (this was easier because I wasn't running anything as intensive as VGA on core0), figuring I'd further optimize using the PIO later.  The approach I took for reading was having a small array of register values so when a read happened the only work to do was use the low address bits to index into this array, push the value out to the Z80 bus, and flip a bit in a status register (which would be reset by core0 code later) so the Z80 could tell if it was reading things too quickly.

I think a similar approach may work here -- probably easier for custom new peripherals than for emulating existing hardware that may have more work that must be done between back-to-back bus operations than cycles available -- but the general model is to try to have the registers always ready to read so you don't have to call out to any other code during the read cycle.

It might not even require writing it in assembly... possibly something like this might be workable:
```c
volatile uint8_t abus_card_register[16];

void __time_critical_func(abus_loop)() {
    while(1) {
        uint32_t value = pio_sm_get_blocking(CONFIG_ABUS_PIO, ABUS_MAIN_SM);
        uint32_t address = (value >> 10) & 0xffff;
        bool is_write = ((value & (1u << (CONFIG_PIN_APPLEBUS_RW - CONFIG_PIN_APPLEBUS_DATA_BASE))) == 0);
        if(((value & (1u << (CONFIG_PIN_APPLEBUS_DEVSEL - CONFIG_PIN_APPLEBUS_DATA_BASE))) == 0)) {
            if (!is_write) {
                // device read access
                pio_sm_put_blocking(CONFIG_ABUS_PIO, ABUS_DEVICE_READ_SM, abus_card_register[address & 0xf]);
            }
            // possibly signal that a read occurred, flip a status bit somewhere, etc...
            gpio_xor_mask(1u << PICO_DEFAULT_LED_PIN);
        }
        shadow_memory(is_write, address, value);
    }
}
```

I'm stuffing another board so I can use one for a display and a second as an experimental peripheral and try this out and see what there is to see.

**Reply by swetland** [reply]
So I tried this out and reads from the Apple IIe were unreliable until I commented out the `shadow_memory(...)` call.  Looks like sometimes the time spent in there is enough to miss the window on responding to a following read.

I think one approach that might work would be to immediately process videoram writes, but queue softswitch writes to be handled during vsync ahead of the next frame.  Though it's also possible that one could tune the `shadow_memory(...)` handling so it's fast enough to always complete in time for the next bus access.

**Reply by swetland** [reply]
Hmm.  You need the softswitch writes processed to handle videoram writes, so deferring them is not going to work.

**Comment by swetland** [comment]
An excessively debug-instrumented Apple II VGA board (SWD debug probe, Serial to USB adapter, and big red physical reset button just in case):

![appleiivga-debug-setup](https://github.com/markadev/AppleII-VGA/assets/94628/2f675953-cfd1-4aa4-81a5-901e444e22de)

We'll see if I can get away without stuffing the headers for connecting an oscilloscope or logic analyzer to the Apple BUS or not...

**Comment by marketideas** [comment]
So once you get this going, you might want to consider dropping the Pico board and going directly to an RP2040...you will get your extra GPIOs and can use 3 of them to add one more bit of color to each RGB. Thinking ahead for the iigs version.

**Comment by swetland** [comment]
Quick question on timing, @markadev  -- I notice you figure an 18ns buffer plus clock input delay for rising PHI0 and 28ns for falling.  Are those based on measured or estimated times from the '07, and/or do they factor in some data on general RP2040 input delays?

While tinkering with the PIO programs to handle IOSEL as well as DEVSEL (I've now got a readable "rom" at 0xCn00 as well as the register bank) I noticed this and realized that the tPHL/tPLH on the 74LVC245As (which I'm using for the clock) are not asymmetrical and also faster (2.9ns typ, 6.3ns max), so I may want to slightly tweak these for my board.  I probably should just put a scope on it and see exactly what's going on, but I'm still curious for the basis of your timing computations.

Thanks!

The experimental scanline effect looks nice. I wired up /DEVSEL 0xF as a toggle so I could flip it on and off from the IIe itself, but I'm thinking I'll probably end up leaving them on.

**Reply by markadev** [reply]
Hey there.  That 18ns figure includes an estimated delay through the the 74LS07 but also a measured delay in the RP2040 inputs.

With my scope I was measuring the time from the buffer's output going high to the time that other Pico outputs (like the transceiver controls) were changed by the PIO program.  Since the PIO timing is so deterministic any unaccounted for time I just called "input delay" and then added in the 74LS07's tPLH (6ns typ/10ns max).  IIRC I made the same estimation on the PHI0 falling edge, but with 74LS07's tPHL (19ns typ/30ns max).

Very cool that you could get both IOSEL and DEVSEL reads working!
